Complete first version of the cluster migration (#2167)
giantswarm/docs: src/content/reference/fleet-management/migration_to_cluster_api.md

---
title: Migration to Cluster API

> description: A explanation on how the migration from our old Vintage management clusters to Cluster API works.

Author: ```suggestion
description: An explanation of how the migration from our old vintage management clusters to Cluster API works.
```

> title: Migration to Cluster API
> description: An explanation of how the migration from our old vintage management clusters to Cluster API works.

Reviewer: ```suggestion
title: Migration to Cluster API
description: How the migration from our old vintage management and workload clusters to Cluster API works.
```

weight: 20
last_review_date: 2024-03-14
owner:
  - https://github.com/orgs/giantswarm/teams/sig-docs
user_questions:
  - What are the requirements for migrating a cluster to Cluster API?
  - What are the recommendations for a smooth migration?
---

> In Giant Swarm we have used Kubernetes to build platforms since the beginning. On the first years there was no standard process provided by the community, and we develop our custom solution to manage the lifecycle of the clusters. It has worked well for us and our customers, but we learnt that many companies in the ecosystem were resolving the same problem over and over again, so we foster and push for joining forces and build a standard approach. [Cluster API]({{< relref "/reference/fleet-management/introduction" >}}) is backed by the Kubernetes community and covers different providers like AWS, Azure, GCP, and others.

Reviewer 2: ```suggestion
From the outset, Giant Swarm has utilized Kubernetes to build platforms. In the early years everybody was still figuring out how to effectively manage Kubernetes Cluster Lifecycles -- as were we. We built our own tooling, largely based on Operators, which worked well for us and our customers. As the Kubernetes Project and the community around it evolved, it became clear that many companies in the ecosystem were trying to solve the same fundamental challenges regarding Cluster Lifecycle Management. With our extensive experience, we saw an opportunity to contribute to a broader solution. We pushed for a joint effort to build a standard approach for Cluster Lifecycle Management. [Cluster API]({{< relref "/reference/fleet-management/introduction" >}}) is backed by the Kubernetes community and covers different providers like AWS, Azure, GCP, and others.
```

> From the outset, Giant Swarm has utilized Kubernetes to build platforms. In the early years, everybody was still figuring out how to effectively manage Kubernetes lifecycle across a fleet of clusters. We built our own tooling, largely based on [operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/), which worked well for us and our customers. As the Kubernetes project and the community around it evolved, it became clear that many companies in the ecosystem were trying to solve the same fundamental challenges regarding cluster lifecycle management. With our extensive experience, we saw an opportunity to contribute to a broader solution. We pushed for a joint effort to build a standardized method for cluster lifecycle management. [Cluster API]({{< relref "/reference/fleet-management/introduction" >}}) is backed by the Kubernetes community and covers different providers like AWS, Azure, GCP, and others.

Reviewer: Very nice intro

> In this document we explain the migration path from our vintage system to the [Cluster API](https://cluster-api.sigs.k8s.io/) standard. The migration process was designed to ensure a smooth transition of customer workload clusters from our vintage generation to the new Cluster API (CAPI) flavour. This document provides an overview of the process, its requirements, and recommendations to ensure a successful migration.

Reviewer 2: ```suggestion
This guide outlines the migration path from our "vintage" platform to the [Cluster API](https://cluster-api.sigs.k8s.io/) (CAPI) standard, ensuring a seamless transition for customers workload clusters from the traditional system to the modern CAPI framework. Within this document, you'll find a comprehensive overview of the migration procedure, including its prerequisites and strategic advice, all aimed at facilitating a smooth and successful transition.```

> This guide outlines the migration path from our vintage platform to the [Cluster API](https://cluster-api.sigs.k8s.io/) (CAPI) standard, ensuring a seamless transition for customers workload clusters from the traditional system to the modern CAPI framework. Within this document, you'll find a comprehensive overview of the migration procedure, including its prerequisites and strategic advice, all aimed at facilitating a smooth and successful transition.```

Reviewer: ```suggestion
This guide outlines the migration path from our vintage platform to the [Cluster API](https://cluster-api.sigs.k8s.io/) (CAPI) standard, ensuring a seamless transition for customer workload clusters from the previous system to the modern CAPI framework. Within this document, you'll find an overview of the migration procedure, including its prerequisites and strategic advice, all aimed at facilitating a smooth and successful transition.
```

Or `customers'`. And I removed "comprehensive" since it's basically the opposite of "overview".

## Pre-migration requirements

Before you begin the migration:

> 1. Your cluster should be on a release version `>=20.0.0`.

Reviewer: This release is AWS-specific, so let's specify that "Vintage AWS" is meant in this article, and link to the release notes

2. An AWS IAM role, named `giantswarm-{CAPI_MC_NAME}-capa-controller`, must be pre-created for the workload cluster's (WC) AWS account. [More information on how to do it in this guide](https://docs.giantswarm.io/vintage/getting-started/cloud-provider-accounts/cluster-api/aws/#overview).

__Note:__ The `CAPI_MC_NAME` is the name of the management cluster (MC) where the Cluster API controllers are installed.

## Recommendations for a smooth migration

In preparation for migration, we recommend increasing the size of your "master" node instance type to two or three times its actual size (example, from `mx.large` to `mx.4xlarge`). This ensures the API server can efficiently handle the load during migration since there might be only one node to handle the traffic at certain points.

## The migration process

The migration process consists of several steps:

> 1. __Initialization:__ Retrieval of necessary Kubernetes access credentials together with AWS credentials. At the same time, it creates a vault client to interact with the Vault instance that contains all security assets of the cluster.

Reviewer 2: ```suggestion
1. __Initialization:__ Necessary Kubernetes access credentials and AWS credentials are retrieved.  Vault client is created to interact with the Vault instance containing all security assets of the cluster.
```

> We also recommend increasing the size of your "master" node instance type to 2x or 3x its normal size for the duration of the migration. (e.g. `mx.large` to `mx.4xlarge`). This ensures the API server can effectively handle the load during the migration since there might only be one node to handle the traffic at certain points throughout the process.

Reviewer: ```suggestion
We also recommend increasing the size of your "master" node instance type to 2x or 3x its normal size for the duration of the migration. (e.g. `m5.large` to `m5.4xlarge`). This ensures the API server can effectively handle the load during the migration since there might only be one node to handle the traffic at certain points throughout the process.
```

> The migration process consists of several steps:

Reviewer: Should we state that _we_ automate and run this?

Also, we haven't stated yet that _a new management cluster_ will be used for the transfer. These high-level things are more relevant to customers than the following details, and should be mentioned pretty early in this article.

Reviewer 3: definitely worth mentioning hat we migrate from one CM to a new one

> 2. __Preparation:__ Migration of secrets to the CAPI management cluster, including CA certs, encryption provider secrets, and service account secrets. The migration scripts are created as a secret in the CAPI management cluster. Additionally, AWS credentials for the cluster are migrated by creating an `AWSClusterRoleIdentity` in the CAPI management cluster. There are certain operations performed to avoid conflicts during migration, such as disabling machine health check on the vintage cluster resources, scaling down the app operator for the migrated workload cluster, or cleaning up certain charts.

Reviewer 2: ```suggestion
2. __Preparation:__ Migration of secrets to the CAPI management cluster, including CA certs, encryption provider secrets, and service account secrets. Migration scripts are created as secrets in the CAPI management cluster. Additionally, AWS credentials for the cluster are migrated by creating an `AWSClusterRoleIdentity` in the CAPI management cluster. Certain operations are performed to avoid conflicts during migration, such as disabling machine health check on the vintage cluster resources, scaling down the app operator for the migrated workload cluster, or cleaning up certain charts.
```

3. __Stopping vintage cluster resource reconciliation:__ To avoid conflicts, vintage reconciliation is stopped by removing all `aws-operator` labels from the vintage cluster resource.
4. __Cluster API cluster provisioning:__ Generation and application of CAPI cluster templates. A separate routine runs in the background to ensure the old load balancer remains active. The tool waits until at least one CAPI control-plane node joins the cluster and is in a Ready state. Various operations are performed to ensure a smooth transition, such as stopping control-plane components on the vintage cluster, cordoning all vintage control-planes, and deleting certain pods to speed up installation and updates.

> 5. __Cleaning the vintage cluster:__ All vintage control-plane nodes are drained, vintage auto scaling groups are deleted, and all worker nodes for each node pool are drained and deleted.

Reviewer: We should mention any expectations the customer can have regarding uptime

Reviewer 3: I think the goal is to have the same level of service as you would expect in a normal `update` so no downtime.

## Post-migration cleanup

> Our engineer will check that all resources and infrastructure are correctly migrated and that the new cluster is working as expected.

Reviewer: ```suggestion
Our engineers will check that all resources and infrastructure are correctly migrated and that the new cluster is working as expected.
```

> By following this process, you can ensure a smooth and efficient migration of your Kubernetes clusters to the Cluster API. If you encounter any issues during the migration process, our team is ready to assist you.

Reviewer: > By following this process, you

Sounds as if the customer runs something. Let's reword to say that it's done together (?).

Reviewer 3: +1

---
title: Migration to Cluster API
description: A explanation on how the migration from our old Vintage management clusters to Cluster API works.
weight: 20
last_review_date: 2024-03-14
owner:
  - https://github.com/orgs/giantswarm/teams/sig-docs
user_questions:
  - What are the requirements for migrating a cluster to Cluster API?
  - What are the recommendations for a smooth migration?
---

> In Giant Swarm we have used Kubernetes to build platforms since our beginning. We started with a custom solution to manage the lifecycle of the clusters due to the lack of a standard way out of there. It has worked well for us and our customers, but we learnt many companies were resolving the same problem over and over again, so we foster and push for joining forces and build a standard approach. [Cluster API]({{< relref "/reference/fleet-management/introduction" >}}) is backed by the Kubernetes community and covers different providers like AWS, Azure, GCP, and others.

Reviewer 4: Could you explain the meaning of `due to lack of a standard way out of there`? I interpret it as `finding a way out of a given situation`, but it's not clear to me what the situation is from the paragraph.

Author: Well in 2015 we created our own solution since there was no CAPI. I will make it clearer

In this document we explain the migration path from our vintage system to the [Cluster API](https://cluster-api.sigs.k8s.io/) standard. This document provides an overview of the process, its requirements, and recommendations to ensure a successful migration.

## Goals

> Our new migration process looks for ensuring a smooth transition of your Kubernetes workload clusters from vintage generation to the new Cluster API (CAPI) flavour. This document provides an overview of the process, its requirements, and recommendations to ensure a successful migration.

Reviewer 4: Maybe the previous paragraph could be integrated into this one because most of the information is duplicated.

## Pre-migration requirements

Before you begin the migration:

1. Your cluster should be on a release version `>=20.0.0`.

> 2. An AWS IAM role, named `giantswarm-{CAPI_MC_NAME}-capa-controller`, must be pre-created for the workload cluster's (WC) AWS account.

Reviewer 4: A link to how to create the role is probably needed.

__Note:__ The `CAPI_MC_NAME` is the name of the management cluster (MC) where the Cluster API controllers are installed.

## Recommendations for a smooth migration

> In preparation for migration, we recommend increasing the size of your "master" node instance type to tow or three times its actual size (example, from `mx.large` to `mx.4xlarge`). This ensures the API server can efficiently handle the load during migration since there might be only one node to handle the traffic at certain points.

Reviewer 4: Typo `tow` should be `two`.

3. __Stopping vintage cluster resource reconciliation:__ To avoid conflicts, vintage reconciliation is stopped by removing all `aws-operator` labels from the vintage cluster resource.
4. __Cluster API cluster provisioning:__ Generation and application of CAPI cluster templates. A separate routine runs in the background to ensure the old load balancer remains active. The tool waits until at least one CAPI control-plane node joins the cluster and is in a `Ready` state. Various operations are performed to ensure a smooth transition, such as stopping control-plane components on the vintage cluster, cordoning all vintage control-planes, and deleting certain pods to speed up installation and updates.
5. __Cleaning the vintage cluster:__ All vintage control-plane nodes are drained, vintage auto scaling groups are deleted, and all worker nodes for each node pool are drained and deleted.

__Note:__ The migration process is automated and executed by our engineers. The process is monitored and controlled by a set of scripts and tools to ensure a smooth transition. We don't expect any downtime for the workload clusters during the migration process.

## Post-migration tasks

> Our engineers will check that all resources and infrastructure are correctly migrated and that the new cluster is working as expected. There are some reminding tasks needed to be done by the customer:

Author: ```suggestion
Our engineers will check that all resources and infrastructure are correctly migrated and that the new cluster is working as expected. There are some remaining tasks that need to be done by the customer:
```

> - The DNS setup changes for the workload clusters. The new management cluster has a new host zone allocated in AWS. In the vintage setup, the host zone contained the workload cluster name meanwhile in the CAPI setup the DNS structure is more flexible not containing the workload cluster name. Both the old and new host zones will be available for a certain period to ensure a smooth transition, but customers should migrate the DNS records to the new host zone as soon as possible in case they are using cluster wildcard DNS records.

Reviewer 3: old scheme contains  MC name, new scheme does not contain MC name, both contain WC name
